components/compliance-service/profiles/db/store.go

```diff
@@ -376,89 +379,108 @@ func (s *Store) Delete(namespace string, name string, version string) error {
-	if order != "ASC" && order != "DESC" {
-		return nil, status.Errorf(codes.InvalidArgument, "order field '%s' is invalid. Use either 'ASC' or 'DESC'", order)
-	}
-
-	if sort != "name" && sort != "title" && sort != "maintainer" {
-		return nil, status.Errorf(codes.InvalidArgument, "sort field '%s' is invalid. Use either 'name', 'title' or 'maintainer'", sort)
-	}
-
-	var rows *sql.Rows
-	var err error
-
-	switch {
-	case len(namespace) == 0 && len(name) == 0:
-		query := fmt.Sprintf(`
-			SELECT sha256, info
-			FROM store_profiles WHERE sha256 IN
-			(SELECT sha256 FROM store_market)
-			ORDER BY store_profiles.info #>> '{%s}' %s, store_profiles.info #>> '{version}' DESC;
-		`, sort, order)
-
-		rows, err = s.DB.Query(query)
-	case len(namespace) == 0:
-		query := fmt.Sprintf(`
-			SELECT sha256, info
-			FROM store_profiles WHERE sha256 IN
-			(SELECT sha256 FROM store_market)
-			AND info->>'name' = $1
-			ORDER BY store_profiles.info #>> '{%s}' %s, store_profiles.info #>> '{version}' DESC;
-		`, sort, order)
-
-		rows, err = s.DB.Query(query, name)
-	case len(name) == 0:
-		query := fmt.Sprintf(`
-			SELECT sha256, info
-			FROM store_profiles WHERE
-			sha256 IN (SELECT sha256 FROM store_namespace WHERE owner=$1)
-			ORDER BY store_profiles.info #>> '{%s}' %s, store_profiles.info #>> '{version}' DESC;
-		`, sort, order)
-
-		rows, err = s.DB.Query(query, namespace)
-	default:
-		query := fmt.Sprintf(`
-			SELECT sha256, info
-			FROM store_profiles WHERE
-			sha256 IN (SELECT sha256 FROM store_namespace WHERE owner=$1)
-			AND info->>'name' = $2
-			ORDER BY store_profiles.info #>> '{%s}' %s, store_profiles.info #>> '{version}' DESC;
-		`, sort, order)
-
-		rows, err = s.DB.Query(query, namespace, name)
+type ProfilesListRequest struct {
+	Filters   map[string][]string
+	Name      string
+	Namespace string
+	Order     string
+	Page      int
+	PerPage   int
+	Sort      string
+}
+
+func (s *Store) ListProfilesMetadata(req ProfilesListRequest) ([]inspec.Metadata, int, error) {
+	if req.Order != "ASC" && req.Order != "DESC" {
+		return nil, 0, status.Errorf(codes.InvalidArgument, "order field '%s' is invalid. Use either 'ASC' or 'DESC'", req.Order)
+	}
+
+	if req.Sort != "name" && req.Sort != "title" && req.Sort != "maintainer" {
+		return nil, 0, status.Errorf(codes.InvalidArgument, "sort field '%s' is invalid. Use either 'name', 'title' or 'maintainer'", req.Sort)
+	}
+
+	sql := squirrel.
+		Select("sha256, info, count(*) OVER() AS total").
+		From("store_profiles")
+
+	if len(req.Namespace) == 0 {
+		sql = sql.
+			Where("exists(select 1 from store_market where store_profiles.sha256 = store_market.sha256)")
+	} else {
+		sql = sql.
+			Where("exists(select 1 from store_namespace where store_profiles.sha256 = store_namespace.sha256 and store_namespace.owner = ?)", req.Namespace)
+	}
+
+	if len(req.Name) != 0 {
+		sql = sql.Where("info->>'name' = ?", req.Name)
```

Comment:
What is the `->>` notation here (and likewise the `#>>` in the original)? Not familiar with that.

Reply:
They are both postgres json operators: https://www.postgresql.org/docs/9.6/functions-json.html

`#>>` is a more powerful operator to get an element at a specified path (hence the `{}` brackets in the original), `->>` is used to get the value at a top-level key. The extra `>` in both of those operators (ie `->>` vs `->`) means it'll always return a string instead of null or a string.